packages/serialize/types/index.d.ts

```diff
@@ -27,18 +32,22 @@ export type InterpolationPrimitive =
   | ComponentSelector
   | Keyframes
   | SerializedStyles
-  | CSSObject
+  | CSSObject<Props>
 
-export type CSSInterpolation = InterpolationPrimitive | ArrayCSSInterpolation
+export type CSSInterpolation<Props = unknown> =
```

Comment:
This one actually shouldn't end up including functions in `CSSObject`. We have 2 flavors of those interpolations:
- the ones that u can pass to `css`
- and the ones that you can pass to `styled` and similar

The difference is that `css` is a context-less call. We don't get access to any `props` there (and we don't defer `css` resolution) so this should be an error:
```ts
css({
  display: () => 'flex'
})
```

That's why we have both `ArrayInterpolation` and `ArrayCSSInterpolation` here. A similar distinction should be introduced for those object styles.

But now... I'm not sure if this would satisfy your original goal.

Reply:
I see, I will revert this. I think it should be fine since we focus only for `styled`.

Reply:
Done!

Reply:
> I think it should be fine since we focus only for styled.

Cool! ❤️ 

> Done! see https://github.com/emotion-js/emotion/commit/2fb22d4dc23a7bc7a81cbd88eacaaf02ea9d1077

Thanks. I'm soft-approving this PR now. I think we still need to tweak types a little bit but honestly - I need to give it more thought to determine how to do it in the least disruptive way. So I think, it's best for me to do it soon-ish since I don't have concrete advice right now on how this should be done.

I'll also give @emmatown a couple of days to give their opinion about this. If we don't merge it this week - please ping me later.